type.md

```diff
@@ -250,13 +253,134 @@ The value of ARG is "Hello", which is not of type NUMBER.
-guarantee that Lisp compiler will perform compile-time type checking.
+guarantee that the Lisp compiler will perform compile-time type checking.
+
+However, it is possible, and SBCL is an implementation that does
+thorough type checking.
+
+### Declaring the type of variables
+
+Use the macro [`declaim`][declaim].
+
+Let's declare that our global variable `*name*` is a string (you can
+type the following in any order in the REPL):
+
+~~~lisp
+(declaim (type (string) *name*))
+(defparameter *name* "book")
+~~~
+
+Now if we try to set it with a bad type, we get a `simple-type-error`:
+
+~~~lisp
+(setf *name* :me)
+Value of :ME in (THE STRING :ME) is :ME, not a STRING.
+   [Condition of type SIMPLE-TYPE-ERROR]
+~~~
+
+We can do the same with our custom types. Let's quickly declare the type `list-of-strings`:
+
+~~~lisp
+(defun list-of-strings-p (list)
+  "Return t if LIST is non nil and contains only strings."
+  (and (consp list)
+       (every #'stringp list)))
+
+(deftype list-of-strings ()
+  `(satisfies list-of-strings-p))
+~~~
+
+Now let's declare that our `*all-names*` variables is a list of strings:
+
+~~~lisp
+(declaim (type (list-of-strings) *all-names*))
+(defparameter *all-names* "")
+~~~
+
+We can compose types:
+
+~~~lisp
+(declaim (type (or null list-of-strings) *all-names*))
+~~~
+
+### Declaring the input and output types of functions
+
+We use again the `declaim` macro, with `ftype (function …)` instead of just `type`:
+
+~~~lisp
+(declaim (ftype (function (fixnum) fixnum) add))
+;;                         ^^input ^^output [optional]
+(defun add (n)
+	(+ n  1))
+~~~
+
+With this we get nice type warnings at compile time.
+
+If we change the function to erroneously return a string instead of a
+fixnum, we get a warning:
+
+~~~lisp
+(defun add (n)
+	(format nil "~a" (+ n  1)))
+; caught WARNING:
+;   Derived type of ((GET-OUTPUT-STREAM-STRING STREAM)) is
+;     (VALUES SIMPLE-STRING &OPTIONAL),
+;   conflicting with the declared function return type
+;     (VALUES FIXNUM &REST T).
+~~~
+
+If we use `add` inside another function, to a place that expects a
+string, we get a warning:
+
+~~~lisp
+(defun bad-concat (n)
+    (concatenate 'string (add n)))
+; caught WARNING:
+;   Derived type of (ADD N) is
+;     (VALUES FIXNUM &REST T),
+;   conflicting with its asserted type
+;     SEQUENCE.
+~~~
+
+If we use `add` inside another function, and that function declares
+its argument types which appear to be incompatible with those of
+`add`, we get a warning:
+
+~~~lisp
+(declaim (ftype (function (string)) bad-arg))
+(defun bad-arg (n)
+    (add n))
+; caught WARNING:
+;   Derived type of N is
+;     (VALUES STRING &OPTIONAL),
+;   conflicting with its asserted type
+;     FIXNUM.
+~~~
+
+This all happens indeed *at compile time*, either in the REPL,
+either with a simple `C-c C-c` in Slime, or when we `load` a file.
+
+
+After this, no need to recall that lisp already warns about simple
```

Comment:
I'm not sure what this question means rhetorically in the context.

Reply:
indeed. I kept the idea, rephrased and moved it at the beginning of the section.